libcontainer/cgroups/fs/cpu.go

```diff
@@ -35,17 +36,70 @@ func (s *CpuGroup) Apply(path string, r *configs.Resources, pid int) error {
 }
 
 func (s *CpuGroup) SetRtSched(path string, r *configs.Resources) error {
-	if r.CpuRtPeriod != 0 {
-		if err := cgroups.WriteFile(path, "cpu.rt_period_us", strconv.FormatUint(r.CpuRtPeriod, 10)); err != nil {
+	const (
+		filePeriod  = "cpu.rt_period_us"
+		fileRuntime = "cpu.rt_runtime_us"
+	)
+
+	if r.CpuRtPeriod == 0 && r.CpuRtRuntime == 0 {
+		return nil
+	}
+
+	if r.CpuRtPeriod != 0 && r.CpuRtRuntime == 0 {
+		return cgroups.WriteFile(path, filePeriod, strconv.FormatUint(r.CpuRtPeriod, 10))
+	} else if r.CpuRtRuntime != 0 && r.CpuRtPeriod == 0 {
+		return cgroups.WriteFile(path, fileRuntime, strconv.FormatInt(r.CpuRtRuntime, 10))
+	}
+
+	// When neither CpuRtPeriod nor CpuRtRuntime is equal to 0, let's set them in the correct order
+
+	sOldPeriod, err := cgroups.ReadFile(path, filePeriod)
+	if err != nil {
+		return err
+	}
+	oldPeriod, err := strconv.ParseInt(strings.TrimSpace(sOldPeriod), 10, 64)
+	if err != nil {
+		return err
+	}
+
+	sOldRuntime, err := cgroups.ReadFile(path, fileRuntime)
+	if err != nil {
+		return err
+	}
+	oldRuntime, err := strconv.ParseInt(strings.TrimSpace(sOldRuntime), 10, 64)
+	if err != nil {
+		return err
+	}
+
+	/*
+		When we set a new rt_period_us, the kernel will determine
+		whether the current configuration of new_limit1 = old_quota/new_period
+		exceeds the limit. If it exceeds the limit, an error will be reported.
+		Maybe it is reasonable to set rt_runtime_us first.
+		new_limit2 = new_quota/old_period.
```

Comment:
This line should be together with the next paragrapha, place it here looks strange.